src/traces/cone/attributes.js

```diff
@@ -80,9 +80,28 @@ var attrs = {
         description: 'Sets the z coordinates of the vector field mesh.'
     },
 
-    // TODO
-//     sizemode: {},
-//     sizeref: {},
+    sizemode: {
+        valType: 'enumerated',
+        values: ['scaled', 'absolute'],
+        role: 'info',
+        editType: 'calc',
+        dflt: 'scaled',
+        description: [
+            'Sets the mode by which the cones are sized.',
+            'If *scaled*, `sizeref` scales such that the reference cone size',
+            'for the maximum vector magnitude is 1.',
+            'If *absolute*, `sizeref` scales such that the reference cone size',
+            'for vector magnitude 1 is one grid unit.'
```

Comment:
To be honest, I'm not sure. I copied this from https://github.com/gl-vis/gl-cone3d#constructor I'll have to inspect  

https://github.com/gl-vis/gl-cone3d/blob/c675b25e5991e63d0ad23ed24d712c496c28cd72/cone.js#L138-L246

a little bit more.

Reply:
I'll have a new look at the cone sizing parameters. Pre-meshgrid they were based on the position space units and scaled so that two cones with maximum norm, pointing to the same direction, would have their tip and base touch. Like <|<|. That would be with coneSize 2. With coneSize 1, you'd get no overlapping cones even when they're |><|.

The absolute sizing ditched the automatic cone scaling and used the cone vector norm to figure out the size of the cone model, so that a norm of 1 with absolute size factor 1 results in a cone that's one unit tall in the position unit space. Absolute size factor 2 => 2 unit tall cone, etc.

With the meshgrid sampling (and the realization that the cone positions may not be on a nice regular grid), I changed the automatic cone sizing to scale based on the smallest distance between two adjacent cone positions (adjacent in the sense that they've got successive indices in the positions array). This is not the right thing to do though, doing a smallest distance between any two cone positions would be, but that's O(n^2) with a naive algo. 

Another idea floated on the sizing was to have pixel-based sizing as well, which'd probably work like "invert the projection, view and model transform matrices for cone model, scale cone model according to viewport size" but would result in orthogonal camera cones which might be confusing. (And it sounds like a bit of a pain to implement.)

Reply:
@kig that all sounds great. I guess we only need to worry about the smallest-distance calculation when using the `(x, y, z)` data to position the cones; With the cartesian product of x/y/z interpolation grids each dimension should be pretty compact, even if the values are out of order. But I guarantee at some point we'll see a data set where the closest points are not successive. I don't know if there's a JS implementation already available but it's possible to do it in O(n log(n)^2) http://www.cs.ucsb.edu/~suri/cs235/ClosestPair.pdf

Lets not worry about pixel-based sizing, at least not for now. I guess I can see using this if we also do automatic grid spacing, ie when you zoom in and the cone spacing gets larger, at some point we decrease the grid size so we draw more cones between the existing ones. That's definitely not a v1 feature 😅